cocos2d/core/base-nodes/CCNodeCanvasRenderCmd.js

```diff
@@ -495,21 +507,57 @@ cc.Node.RenderCmd.prototype.originTransform = cc.Node.RenderCmd.prototype.transf
 
 //-----------------------Canvas ---------------------------
 
-(function() {
+(function () {
 //The cc.Node's render command for Canvas
     cc.Node.CanvasRenderCmd = function (renderable) {
         cc.Node.RenderCmd.call(this, renderable);
         this._cachedParent = null;
         this._cacheDirty = false;
+        this._currentRegion = new cc.Region();
+        this._oldRegion = new cc.Region();
+        this._regionFlag = 0;
+        this._canUseDirtyRegion = false;
+    };
+
+    cc.Node.CanvasRenderCmd.RegionStatus = {
+        NotDirty: 0,    //the region is not dirty
+        Dirty: 1,       //the region is dirty, because of color, opacity or context
+        DirtyDouble: 2  //the region is moved, the old and the new one need considered when rendering
     };
 
     var proto = cc.Node.CanvasRenderCmd.prototype = Object.create(cc.Node.RenderCmd.prototype);
     proto.constructor = cc.Node.CanvasRenderCmd;
 
+    proto._notifyRegionStatus = function (status) {
+        if (this._needDraw && this._regionFlag < status) {
+            this._regionFlag = status;
+        }
+    };
+
+    var localBB = new cc.Rect();
+    proto.getLocalBB = function () {
+        var node = this._node;
+        localBB.x = localBB.y = 0;
+        localBB.width = node._getWidth();
+        localBB.height = node._getHeight();
+        return localBB;
+    };
+
+    proto._updateCurrentRegions = function () {
+        var temp = this._currentRegion;
+        this._currentRegion = this._oldRegion;
+        this._oldRegion = temp;
+        //hittest will call the transform, and set region flag to DirtyDouble, and the changes need to be considered for rendering
+        if (cc.Node.CanvasRenderCmd.RegionStatus.DirtyDouble === this._regionFlag && (!this._currentRegion.isEmpty())) {
```

Comment:
If we can directly use property to check, it's preferable than using function call